mlflow/pyfunc/backend.py

```diff
@@ -137,33 +230,125 @@ def copy_model_into_container(dockerfile_context_dir):
+    s = requirements_file_contents + python_env_file_contents + (env_id or "")
+    return "mlflow-" + hashlib.sha1(s.encode("utf-8")).hexdigest()
+
+
+def _execute_in_virtualenv(local_model_path, command, install_mlflow, env_id, command_env):
+    # Read required metadata
+    python_env_file = os.path.join(local_model_path, _PYTHON_ENV_FILE_NAME)
+    python_env = _PythonEnv.from_json(python_env_file)
+
+    before = time.time()
+    _logger.info("Start creating environment")
+
+    # Install python
+    _logger.info("Installing python %s", python_env.python_version)
+    is_windows = _is_windows()
+    # pyenv-win automatically skips the installation
+    pyenv_install_options = () if is_windows else ("--skip-existing",)
+    # pyenv_install_options = ("--force",)
+    _run_command(
+        ["pyenv", "install", *pyenv_install_options, python_env.python_version], command_env
+    )
+
+    # Get the downloaded python location
+    with tempfile.TemporaryDirectory() as tmpdir:
+        # Run `pyenv local` in a temporary directory to avoid creating a `.pyenv-version` file
+        # in the current working directory.
+        python_exec_path = _get_output(
+            [
+                *_get_entrypoint(),
+                _join_commands(
+                    f"pyenv local {python_env.python_version}",
+                    "pyenv which python",
+                ),
+            ],
+            cwd=tmpdir,
+        ).strip()
+    _logger.info("Python executable path: %s", python_exec_path)
+
+    # Generate an environment name
+    home = os.path.expanduser("~")
+    requirements_file = os.path.join(local_model_path, _REQUIREMENTS_FILE_NAME)
+    env_name = _get_env_name_for_virtualenv(requirements_file, python_env_file, env_id)
+
+    # Create an environment
+    env_root = os.path.join(home, ".mlflow", "envs")
+    os.makedirs(env_root, exist_ok=True)
+    env_dir = os.path.join(env_root, env_name)
+    env_already_exists = os.path.exists(env_dir)
+    if not env_already_exists:
+        _logger.info("Creating a new environment: %s", env_name)
+        _run_command(
+            [
+                "virtualenv",
+                "--python",
+                python_exec_path,
+                *python_env.to_virtualenv_options(),
+                env_dir,
+            ],
+            command_env,
         )
+    else:
+        _logger.info("Environment %s already exists", env_name)
+
+    # Construct a command to activate the environment
+    paths = ("Scripts", "activate.bat") if is_windows else ("bin", "activate")
+    activator = os.path.join(env_dir, *paths)
+    activate_cmd = activator if is_windows else f"source {activator}"
+
+    # Install pip dependencies
+    if not env_already_exists:
+        _logger.info("Installing pip dependencies")
+        pip_install = f"pip install -r {requirements_file}"
+        commands = [activate_cmd, pip_install]
+        _run_command([*_get_entrypoint(), _join_commands(*commands)], command_env)
```

Comment:
Suppose here raise some exception (e.g. because of network issue), then we need to wrap it a `try ... finally` and in `finally` delete the virutalenv env dir.
Otherwise the env dir is a broken environment but next time when `_execute_in_virtualenv` called the broken env dir will be reused but it is still broken.

Reply:
I guess conda side might have similar issue. If it has, let's do similar fix.

Reply:
Good suggestion, I'll make virtualenv environment creation transactional. I think conda should be addressed in a different PR.

Reply:
If our conda env creation logic has a similar issue, I think we should create a separate PR.

Reply:
I tested on conda case:

Command:
```
conda env create -n myenv2  -f .../path/to/conda2.yaml
```

and when the conda creating command running into step "Installing pip dependencies:....", I cut down network, then the conda command failed,
but, then I found the "myenv2" directory has been created, but the created env lack of some package requirements.

So I think we also need to fix this issue on conda side.

Reply:
Let's address this issue in a separate PR.